outdated-pkgs

```diff
@@ -0,0 +1,359 @@
+    print ("<html>\n<head>\n<style type='text/css'>\n.vdiff {color:green}\n"
+           ".rdiff {color:blue}\n</style>\n</head>\n<body>\n<pre>")
+
+if preamble:
+    print preamble
+    print
+
+pkg_diffs = []
+
+if MODE == 'goc':
+    goc_tag = "goc-el%s-%s" % (EL, GOC_REPO)
+    osg_tag = "osg-%s-el%s-%s" % (SERIES, EL, OSG_REPO)
+
+    goc_map = nvrmap(get_koji_list(goc_tag))
+    osg_map = nvrmap(get_koji_list(osg_tag))
+
+    for pkg in sorted(goc_map):
+        if pkg in osg_map:
+            vrcmp = rpmvercmp(goc_map[pkg], osg_map[pkg])
+            if vrcmp < 0 or show_all:
+                pkg_diffs.append([pkg, goc_map[pkg], osg_map[pkg]])
+
+    if pkg_diffs:
+        pkg_diffs[:0] = [["Package", "EL%s GOC" % EL, "OSG " + SERIES]]
+        widths = [ max(map(len,col)) for col in zip(*pkg_diffs) ]
+        pkg_diffs[1:1] = [[ '-' * n for n in widths ]]
+        for i,row in enumerate(pkg_diffs):
+            spacing = [ w-len(x) for x,w in zip(row,widths) ]
+            if use_color and i > 1:
+                row[1:] = colorize_vr(*row[1:])
+            print '  '.join( r + ' ' * s for r,s in zip(row,spacing) ).rstrip()
+    else:
+        print "No package version differences"
+elif MODE == 'osg-epel':
+    epel_map = nvrmap(get_epel_list())
+    osg_map  = nvrmap(get_osg_list())
+    for pkg in sorted(osg_map):
+        if pkg in epel_map:
+            vrcmp = rpmvercmp(osg_map[pkg], epel_map[pkg])
+            if vrcmp < 0 or show_all:
+                pkg_diffs.append([pkg, osg_map[pkg], epel_map[pkg]])
+
+    if pkg_diffs:
+        pkg_diffs[:0] = [["Package", "OSG " + SERIES, "EPEL " + EL]]
+        widths = [ max(map(len,col)) for col in zip(*pkg_diffs) ]
+        pkg_diffs[1:1] = [[ '-' * n for n in widths ]]
+        for i,row in enumerate(pkg_diffs):
+            spacing = [ w-len(x) for x,w in zip(row,widths) ]
+            if use_color and i > 1:
+                row[1:] = colorize_vr(*row[1:])
+            print '  '.join( r + ' ' * s for r,s in zip(row,spacing) ).rstrip()
+    else:
+        print "No package version differences"
+elif MODE == 'osg-i2':
+    i2_map = nvrmap(get_i2_list())
+    osg_map  = nvrmap(get_osg_list())
+
+    for pkg in sorted(osg_map):
+        if pkg in i2_map:
+            vrcmp = rpmvercmp(osg_map[pkg], i2_map[pkg])
+            if vrcmp < 0 or show_all:
+                pkg_diffs.append([pkg, osg_map[pkg], i2_map[pkg]])
+
+    if pkg_diffs:
+        pkg_diffs[:0] = [["Package", "OSG " + SERIES, "I2 EL" + EL]]
+        widths = [ max(map(len,col)) for col in zip(*pkg_diffs) ]
+        pkg_diffs[1:1] = [[ '-' * n for n in widths ]]
+        for i,row in enumerate(pkg_diffs):
+            spacing = [ w-len(x) for x,w in zip(row,widths) ]
+            if use_color and i > 1:
+                row[1:] = colorize_vr(*row[1:])
+            print '  '.join( r + ' ' * s for r,s in zip(row,spacing) ).rstrip()
+    else:
+        print "No package version differences"
```

Comment:
Yeah as @edquist is saying, this looks like a lot of duplicated logic that can be refactored. Similarly for the argparsing above, e.g. each of the "modes" has the following

```
        elif arg == '--show-all':
            show_all = True
        else:
            usage()
```

So you could take them out of the mode `if` blocks and just check them once.

Reply:
I get the idea now. Just pushed two commits that generalized repeated functions.